src/ellswift.rs

```diff
@@ -0,0 +1,661 @@
+    }
+    /// Computes the ElligatorSwift encoding for a valid public key
+    /// # Example
+    /// ```
+    /// use secp256k1::{ellswift::ElligatorSwift, PublicKey, Secp256k1, SecretKey};
+    /// let secp = Secp256k1::new();
+    /// let sk = SecretKey::from_slice(&[1; 32]).unwrap();
+    /// let pk = PublicKey::from_secret_key(&secp, &sk);
+    /// let es = ElligatorSwift::from_pubkey(pk);
+    /// ```
+    pub fn from_pubkey(pk: PublicKey) -> ElligatorSwift { Self::ellswift_encode(pk) }
+    /// Computes a shared secret only known by Alice and Bob. This is obtained by computing
+    /// the x-only Elliptic Curve Diffie-Hellman (ECDH) shared secret between Alice and Bob.
+    /// # Example
+    /// ```
+    /// use secp256k1::{
+    ///     ellswift::{ElligatorSwift, ElligatorSwiftParty},
+    ///     PublicKey, SecretKey, XOnlyPublicKey, Secp256k1,
+    /// };
+    /// use core::str::FromStr;
+    ///
+    /// let secp = Secp256k1::new();
+    ///
+    /// let alice_sk = SecretKey::from_str("e714e76bdd67ad9f495683c37934148f4efc25ce3f01652c8a906498339e1f3a").unwrap();
+    /// let bob_sk = SecretKey::from_str("b6c4b0e2f8c4359caf356a618cd1649d18790a1d67f7c2d1e4760e04c785db4f").unwrap();
+    ///
+    /// let alice_es = ElligatorSwift::from_seckey(&secp, alice_sk, None);
+    /// let bob_es = ElligatorSwift::from_seckey(&secp, bob_sk, None);
+    ///
+    /// let alice_shared_secret = ElligatorSwift::shared_secret(alice_es, bob_es, alice_sk, ElligatorSwiftParty::A, None);
+    /// let bob_shared_secret = ElligatorSwift::shared_secret(alice_es, bob_es, bob_sk, ElligatorSwiftParty::B, None);
+    ///
+    /// assert_eq!(alice_shared_secret, bob_shared_secret);
+    /// ```
+    pub fn shared_secret(
+        ellswift_a: ElligatorSwift,
+        ellswift_b: ElligatorSwift,
+        secret_key: SecretKey,
+        party: ElligatorSwiftParty,
+        data: Option<Vec<u8>>,
+    ) -> ElligatorSwiftSharedSecret {
+        let mut shared_secret = [0u8; 32];
+        unsafe {
+            secp256k1_ellswift_xdh(
+                ffi::secp256k1_context_no_precomp,
+                shared_secret.as_mut_c_ptr(),
+                ellswift_a.0.as_c_ptr(),
+                ellswift_b.0.as_c_ptr(),
+                secret_key.as_c_ptr(),
+                party.into(),
+                ffi::secp256k1_ellswift_xdh_hash_function_bip324,
+                data.map(|d| d.as_ptr()).unwrap_or(std::ptr::null()) as *mut c_void,
+            );
+        }
+        ElligatorSwiftSharedSecret(shared_secret)
+    }
+    /// Constructs an ElligatorSwift object from a 64-byte hex string.
+    /// # Example
+    /// ```
+    /// use secp256k1::ellswift::ElligatorSwift;
+    /// let es = ElligatorSwift::from_hex("c01204f76d2f886a4d3319e2ac0c86a2cb560ec29c732336a99ff152");
+    /// ```
+    pub fn from_hex(hex: &str) -> ElligatorSwift {
+        let mut ser = [0u8; 64];
+        from_hex(hex, &mut ser).unwrap();
+        ElligatorSwift::from_array(ser)
+    }
```

Comment:
In 6cde4c32dad779e69cb2242ef46fe0646e22d806:

We want a `FromStr` and `Display` impl here which use `from_hex`. (I'm ambivalent as to whether `from_hex` itself should be exposed. I'm inclined to say no, and just use `from_str`.)

Also this function definitely needs to return a result and not panic when given bad hex :).